File: python/5. Longest Palindromic Substring.py
# https://leetcode.com/problems/longest-palindromic-substring/

class Solution:
    def longestPalindrome(self, s: str) -> str:
        if len(s) <= 1:
            return s

        def expand(left: int, right: int) -> str :
            while left >= 0 and right < len(s) and s[left] == s[right]:
                left -= 1
                right += 1
            return s[left+1:right]  # 파이썬 슬라이싱은 끝 인덱스 제외
        
        longest = ""

        for i in range(len(s)):
            p1 = expand(i, i)
            p2 = expand(i, i+1)
            longest = max(longest, p1, p2, key=len) # key=len을 쓰면 길이 기준으로 비교 가능
        
        return longest

File: swift/5-Longest-Palindromic-Substring.swift
// https://leetcode.com/problems/longest-palindromic-substring/

class Solution {
    func longestPalindrome(_ s: String) -> String {
        var chars = Array(s)
        var start: Int = 0
        var maxLength: Int = 0

        func expandMaxLength(_ left: Int, _ right: Int) -> Int {
            var l = left
            var r = right

            while l >= 0 && r <= chars.count - 1 && chars[l] == chars[r] {
                l -= 1
                r += 1
            }

            return r - l - 1
        }

        for i in 0..<chars.count {
            let length1 = expandMaxLength(i, i)
            let length2 = expandMaxLength(i, i+1)

            let length = max(length1, length2)
            if maxLength < length {
                start = i - (length - 1) / 2
                maxLength = length
            }
        }

        return String(chars[start..<(start+maxLength)])
    }
}